File: sep2022/minHeap.py
'''
Input: trips = [[2,1,5],[3,3,7]], capacity = 4
Output: false

Input: trips = [[2,1,5],[3,3,7]], capacity = 5
Output: true

Constraints: 0 <= fromi < toi <= 1000

'''


class Solution:
    # O(n)
    def carPooling(self, trips, capacity):
        delta_n = [0] * 1001

        for t in trips:
            num, start, end = t
            delta_n[start] += num
            delta_n[end] -= num

        curr = 0
        for i in range(1001):
            curr += delta_n[i]
            if curr > capacity:
                return False
        return True

    # O(nlogn)
    def carPooling1(self, trips, capacity):
        trips.sort(key=lambda t: t[1])
        minHeap = [] # [end, num]
        curr = 0
        for t in trips:
            num, start, end = t
            while minHeap and minHeap[0][0] <= start:
                curr -= minHeap[0][1]
                heapq.heappop(minHeap)

            curr += num
            if curr > capacity:
                return False
            heapq.heappush(minHeap, [end, num])
        return True

    def carPooling2(self, trips, capacity):
        end = 0
        for i in range(len(trips)):
            if end < trips[i][2]:
                end = trips[i][2]
        for j in range(end):
            c = 0
            for i in range(len(trips)):
                if (trips[i][1] <= j <= trips[i][2]):
                    c += trips[i][0]
            if c > capacity:
                return False
        return True


s = Solution()
print(s.carPooling([[2,1,5],[3,3,7]], 5))
print(s.carPooling1([[2,1,5],[3,3,7]], 5))
print(s.carPooling2([[2,1,5],[3,3,7]], 5))













# end

File: sep2022/minHeap.cpp
# include <iostream>
# include <vector>

using namespace std;

/*
[num, start, end]
[[2,1,5],[3,3,7]]
4 false
5 true
sort ascending
for n < m in range(len(trips))
if trip[n].end > trip[m].start
  trip[n].num + trip[m].num > capacity return false
*/

class Solution {
public:
  bool carPooling(vector<vector<int> > &trips, int capacity) {
    // static array
    int deltaNum[1001];
    // memory memset(ptr, val, num)
    // bytes int sizeof 4
    // memset (str,'-',6); sizeof 1
    memset(deltaNum, 0, 1001*sizeof(int));
    // cout << sizeof(int) << endl;

    for (vector<vector<int> >::iterator it=trips.begin(); it != trips.end(); it++)
    {
      int num = (*it)[0];
      int start = (*it)[1];
      int end = (*it)[2];
      // [[2,1,5],[3,3,7]]
      deltaNum[start] += num;
      deltaNum[end] -= num;
    }
    int curr = 0;
    for (int i=0; i<=1000; i++)
    {
      curr += deltaNum[i];
      if (curr > capacity)
      {
        // cout << curr << " " << i << endl;
        return false;
      }
    }
    return true;
  }

  static bool comp(vector<int> i1, vector<int> i2) {
    return i1[1] < i2[1];
  }
  bool carPooling1(vector<vector<int> > &trips, int capacity) {
    sort(trips.begin(), trips.end(), comp);
    // for (auto i : trips)
    // {
    //   cout << i[1] << endl;
    // }

    vector<vector<int> > minHeap;
    int curr = 0;
    for (vector<vector<int> >::iterator it=trips.begin(); it != trips.end(); it++)
    {
      int num = (*it)[0];
      int start = (*it)[1];
      int end = (*it)[2];

      // minHeap[end, num]
      // keep adding num, as soon as the trip ends, drop off num
      while (minHeap.size()>0 && minHeap[0][0] <= start)
      {
        curr -= minHeap[0][1];
        minHeap.erase(minHeap.begin());
      }

      curr += num;
      if (curr > capacity)
        return false;
      minHeap.push_back(vector<int> {end, num});
    }
    return true;
  }

  // & reference, similiar to pointer, not a copy
  bool carPooling2(vector<vector<int> > &trips, int capacity) {
    // find the largest end stop
    int end = 0;
    for (int i=0; i<trips.size(); i++)
    {
      if (trips[i][2] > end)
      {
        end = trips[i][2];
      }
    }
    for (int j=0; j<end; j++)
    {
      // set to 0 for each j
      int curr = 0;
      for (int i=0; i<trips.size(); i++)
      {
        if (j >= trips[i][1] && j <= trips[i][2])
        {
          curr += trips[i][0];
        }
        if (curr > capacity)
          return false;
      }

    }
    return true;
  }
};

int main(int argc, char* argv[]) {
  vector<int> trips1{2,1,5};
  vector<int> trips2{3,3,7};
  vector<vector<int> > trips{trips1, trips2};
  int capacity = 4;
  Solution sol;
  bool res = sol.carPooling(trips, capacity);
  bool res1 = sol.carPooling1(trips, capacity);
  bool res2= sol.carPooling2(trips, capacity);
  cout << (res ? "True": "False") << endl;
  cout << (res1 ? "True": "False") << endl;
  cout << (res2 ? "True": "False") << endl;


}



// end
